### PR TITLE
Optimise memory consumption, CPU usage and disk writes

### PR DESCRIPTION
**Pull request by nikita-tkachenko-datadog**
### Requirements for Contributing to this repository

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix one issue at the time.
* The pull request must update the test suite to demonstrate the changed functionality.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see [CONTRIBUTING](/CONTRIBUTING.md). 

### What does this PR do?

This PR optimises the plugin's resource consumption:
* heap memory
* CPU usage
* disk I/O

The motivation is that there are several user complaints stating that the overhead of the plugin is too high.

The overhead is mostly caused by how Jenkins' durability mechanism interacts with the plugin. By default Jenkins saves to disk the state of the build at every step (e.g. after finishing a pipeline stage) or with every change (e.g. after adding or removing an `Action` to the build or one of the build's steps). This is needed to ensure durability: if the master node dies, it can be restarted, and the job will resume execution because its up-to-date state will be deserialised from disk.

The problem is that the state saved to disk includes all the data that the plugin associates with the build. This data is added to the build as it executes and is retained until the very end of the build execution. It includes steps/stages execution data, build metadata, etc. The reason it is retained until the end of the build is because the trace for the entire build is submitted once the build is finished.

The problem is aggravated by the fact that the data for individual stages and steps is stored in the build object, so with every step and every change all of that data has to be saved. This results in writing and re-writing to disk multiple times the same data, even the data for stages that have already completed.

This is how it contributes to the overhead:
* heap memory - for pipelines that take a lot of time to finish, all of their data sits in the heap until the very end of the build, increasing pressure on the GC (made worse by the fact that the data from long-living pipelines is likely to survive multiple garbage collections and end up being promoted to old generation).
* CPU usage - the data has to be serialised. By default Jenkins uses reflection-based serialisers that have to examine class metadata in order to determine which fields to serialise and how. This is very CPU-intensive
* disk I/O - as explained above, the data for the entire build is written and re-written to disk many times. Besides, the data contains a lot of duplication, e.g. every step includes its own copy of the pipeline's environment variables.

The points above have been corroborated with several JFR profiles obtained from different customers.

To address the issue, the following is changed:
* As soon as a stage or a step in a pipeline finishes, its span is serialised and added to the next batch that will be submitted. Once this is done, all of the data associated with that stage/step is removed.
* Stage/step-specific data is saved in corresponding build nodes, rather than in the build object. As the result, it is serialised to disk only when there are changes related to those specific build nodes. Unrelated changes to the build or other build nodes will not cause that data to be serialised.
* Custom converters implemented for all data that is serialised. They are more performant than the standard reflection-based converters since there is no need to examine class metadata to determine what needs to be serialised / deserialised.

Since now the plugin retains steps/stages data for the shortest possible period, the way data is propagated from steps to stages to builds (this includes, for example, execution node or git metadata) has been reworked as well.

### Verification Process

Since the changes in the behaviour were minimal, existing tests were used to verify there are no regressions.
Some of the tests had to be adjusted because of the way Git metadata is gathered now: the tests have to do actual SCM checkout to better emulate real-life pipelines.

In addition, manual tests were executed in a dockerized Jenkins instance, covering the following:

Freestyle build trace submitted via webhooks
Freestyle build trace submitted via EVP proxy
Freestyle build trace submitted via APM track
Pipeline trace submitted via webhooks
Pipeline trace submitted via EVP proxy
Pipeline trace submitted via APM track

### Review checklist (to be filled by reviewers)

- [ ] Feature or bug fix MUST have appropriate tests (unit, integration, etc...)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have one `changelog/` label attached. If applicable it should have the `backward-incompatible` label attached.
- [ ] PR should not have `do-not-merge/` label attached.
- [ ] If Applicable, issue must have `kind/` and `severity/` labels attached at least.

### CODE REVIEW

File: src/main/java/org/datadog/jenkins/plugins/datadog/util/git/RepositoryInfoCallback.java [hunk 1]
```diff
@@ -24,25 +26,52 @@ public final class RepositoryInfoCallback implements RepositoryCallback<Reposito
...
-                return RepositoryInfo.EMPTY_REPOSITORY_INFO;
-            }
-
-            final String defaultBranch = GitInfoUtils.normalizeBranch(head.getTarget().getName());
-            return new RepositoryInfo(defaultBranch);
+            String remoteName = getRemoteName(repository);
+            String repoUrl = getRepoUrl(repository, remoteName);
+            final String defaultBranch = getDefaultBranch(repository, remoteName);
+            String branch = GitInfoUtils.normalizeBranch(repository.getBranch()); // currently checked out branch
+            return new RepositoryInfo(repoUrl, defaultBranch, branch);
```
Comment (drodriguezhdez) [inline src/main/java/org/datadog/jenkins/plugins/datadog/util/git/RepositoryInfoCallback.java:33]: If some of this attributes are `null`, I guess we're prepared for it, right?

**Review by drodriguezhdez — COMMENTED** [review]
Dropped some minor comments. Huge refactor, nice work!

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 2]
```diff
@@ -638,35 +605,56 @@ public static Boolean isValidHostname(String hostname) {
...
-        // Check compliance with RFC 1123
-        Pattern r = Pattern.compile(VALID_HOSTNAME_RFC_1123_PATTERN);
-        Matcher m = r.matcher(hostname);
-
         // Final check: Hostname matches RFC1123?
+        Matcher m = VALID_HOSTNAME_RFC_1123_PATTERN.matcher(hostname);
         return m.find();
     }
 
+    private static boolean isPrivateIPv4Address(String ipAddress) {
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:631]: Good point, added

File: src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java [hunk 3]
```diff
@@ -963,28 +946,18 @@ public static String toJson(final Map<String, String> map) {
...
-    /**
-     * Check if a run is from a Jenkins pipeline.
-     * This action is added if the run is based on FlowNodes.
-     *
-     * @param run the current run.
-     * @return true if is a Jenkins pipeline.
-     */
-    public static boolean isPipeline(final Run<?, ?> run) {
-        return run != null && run.getAction(IsPipelineAction.class) != null;
+    public static void cleanUpTraceActions(FlowNode flowNode) {
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java:955]: Makes sense, merged two methods

File: src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogGraphListener.java [hunk 4]
```diff
@@ -167,253 +146,126 @@ public void onNewHead(FlowNode flowNode) {
         }
     }
 
-    @Nullable
-    private BuildPipelineNode buildPipelineNode(FlowNode flowNode) {
-        long start = System.currentTimeMillis();
+    private void processNode(WorkflowRun run, FlowNode flowNode) {
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogGraphListener.java:149]: Added comments here and in some other relevant parts of the code.

File: src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildData.java [hunk 5]
```diff
@@ -342,100 +347,58 @@ private void populateEnvVariables(EnvVars envVars){
...
      */
-    private void populateGitVariables(Run<?,?> run, TaskListener listener, EnvVars envVars) {
-        // First we obtain the actions to check if the Git information was already calculated.
-        // If so, we want to use this information to avoid creating a new Git client instance
-        // to calculate the same information.
-
-        boolean commitInfoAlreadyCreated = isCommitInfoAlreadyCreated(run, this.gitCommit);
-        boolean repositoryInfoAlreadyCreated = isRepositoryInfoAlreadyCreated(run, this.gitUrl);
-
+    private void populateGitVariables(Run<?,?> run) {
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/model/BuildData.java:396]: Added

File: src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogStepListener.java [hunk 6]
```diff
@@ -74,21 +105,217 @@ public void notifyOfNewStep(@Nonnull Step step, @Nonnull StepContext context) {
...
+        if (buildSpanAction == null) {
+            return;
+        }
+        String buildUrl = envVars.get("BUILD_URL");
+        if (buildUrl != null) {
+            buildSpanAction.setBuildUrl(buildUrl);
+        }
+    }
+
+    private static void updateGitData(Run<?, ?> run, Map<String, String> envVars) {
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogStepListener.java:239]: Added

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogBasePipelineLogic.java [hunk 7]
```diff
@@ -82,64 +51,30 @@ protected Set<String> getNodeLabels(Run run, BuildPipelineNode current, String n
...
         }
-
-        return buildData.getNodeName("");
+        // It seems like "built-in" node as the default value does not have much practical sense.
+        // It is done to preserve existing behavior (note that this logic is not applied to metrics - also to preserve the plugin's existing behavior).
+        // The mechanism before the changes was the following:
+        // - DatadogBuildListener#onInitialize created a BuildData instance
+        // - that BuildData had its nodeName populated from environment variables obtained from Run
+        // - the instance was persisted in an Action attached to Run, and was used to populate the node name of the pipeline span (always as the last fallback)
+        // For pipelines, the environment variables that Run#getEnvironment returns _at the beginning of the run_ always (!) contain NODE_NAME = "built-in"
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogBasePipelineLogic.java:64]: The underscores were to emphasise that `NODE_NAME = "built-in"` was only included in env vars at the start of the build. Removed underscored, and added an explicit explanation.

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTraceBuildLogic.java [hunk 8]
```diff
@@ -80,44 +77,46 @@ public TraceSpan createSpan(final BuildData buildData, final Run<?,?> run) {
...
+            buildSpan.putMeta(CITags.CI_PARAMETERS, DatadogUtilities.toJson(buildData.getBuildParameters()));
+        }
+
+        // It seems like "built-in" node as the default value does not have much practical sense.
+        // It is done to preserve existing behavior (note that this logic is not applied to metrics - also to preserve the plugin's existing behavior).
+        // The mechanism before the changes was the following:
+        // - DatadogBuildListener#onInitialize created a BuildData instance
+        // - that BuildData had its nodeName populated from environment variables obtained from Run
+        // - the instance was persisted in an Action attached to Run, and was used to populate the node name of the pipeline span (always as the last fallback)
+        // For pipelines, the environment variables that Run#getEnvironment returns _at the beginning of the run_ always (!) contain NODE_NAME = "built-in"
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTraceBuildLogic.java:93]: The underscores were to emphasise that `NODE_NAME = "built-in"` was only included in env vars at the start of the build. Removed underscored, and added an explicit explanation.

File: src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTraceBuildLogic.java [hunk 9]
```diff
@@ -80,44 +77,46 @@ public TraceSpan createSpan(final BuildData buildData, final Run<?,?> run) {
...
+        // This is true regardless of whether the pipeline definition has a top-level agent block or not
+        // For freestyle projects the correct NODE_NAME seems to be available in the run's environment variables at every stage of the build
+        String nodeName = buildData.getNodeName("built-in");
+        buildSpan.putMeta(CITags.WORKSPACE_PATH, buildData.getWorkspace(""));
         buildSpan.putMeta(CITags.NODE_NAME, nodeName);
 
-        final String nodeLabelsJson = toJson(getNodeLabels(run, nodeName));
-        if(!nodeLabelsJson.isEmpty()){
+        final String nodeLabelsJson = DatadogUtilities.toJson(getNodeLabels(run, buildData.getNodeName("")));
```
[the remark this answers is not in the document]
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/traces/DatadogTraceBuildLogic.java:100]: Added.

File: src/main/java/org/datadog/jenkins/plugins/datadog/util/git/RepositoryInfoCallback.java (hunk 1 above)
Comment (nikita-tkachenko-datadog) [reply src/main/java/org/datadog/jenkins/plugins/datadog/util/git/RepositoryInfoCallback.java:33]: Yes, nothing will fail. I annotated the getters with `@Nullable`, so we'll get a warning if there's a danger of NPE.

**Review by drodriguezhdez — DISMISSED** [review]

**Review by drodriguezhdez — APPROVED** [review]

**Review by drodriguezhdez — DISMISSED** [review]

**Review by drodriguezhdez — APPROVED** [review]

**Pull request merged**